src/client/shadows/dynamicshadowsrender.cpp

```diff
@@ -0,0 +1,563 @@
+		m_driver->removeTexture(shadowMapTextureDynamicObjects);
+
+	if (shadowMapTextureFinal)
+		m_driver->removeTexture(shadowMapTextureFinal);
+
+	if (shadowMapTextureColors)
+		m_driver->removeTexture(shadowMapTextureColors);
+
+	if (shadowMapClientMap)
+		m_driver->removeTexture(shadowMapClientMap);
+}
+
+void ShadowRenderer::initialize()
+{
+	auto *gpu = m_driver->getGPUProgrammingServices();
+
+	// we need glsl
+	if (m_shadows_enabled && gpu && m_driver->queryFeature(video::EVDF_ARB_GLSL)) {
+		createShaders();
+	} else {
+		m_shadows_enabled = false;
+
+		warningstream << "Shadows: GLSL Shader not supported on this system."
+			      << std::endl;
+		return;
+	}
+
+	m_texture_format = m_shadow_map_texture_32bit ? video::ECOLOR_FORMAT::ECF_R32F
+						      : video::ECOLOR_FORMAT::ECF_R16F;
+
+	m_texture_format_color = m_shadow_map_texture_32bit
+						 ? video::ECOLOR_FORMAT::ECF_G32R32F
+						 : video::ECOLOR_FORMAT::ECF_G16R16F;
+}
+
+float ShadowRenderer::getUpdateDelta() const
+{
+	return m_update_delta;
+}
+
+size_t ShadowRenderer::addDirectionalLight()
+{
+	m_light_list.emplace_back(m_shadow_map_texture_size, v3f(0.f, 0.f, 0.f),
+			video::SColor(255, 255, 255, 255), m_shadow_map_max_distance);
+	return m_light_list.size() - 1;
+}
+
+DirectionalLight &ShadowRenderer::getDirectionalLight(u32 index)
+{
+	return m_light_list[index];
+}
+
+size_t ShadowRenderer::getDirectionalLightCount() const
+{
+	return m_light_list.size();
+}
+
+f32 ShadowRenderer::getMaxShadowFar() const
+{
+	if (!m_light_list.empty()) {
+		float wanted_range = m_client->getEnv().getClientMap().getWantedRange();
+
+		float zMax = m_light_list[0].getMaxFarValue() > wanted_range
+					     ? wanted_range
+					     : m_light_list[0].getMaxFarValue();
+		return zMax * MAP_BLOCKSIZE;
+	}
+	return 0.0f;
+}
+
+void ShadowRenderer::addNodeToShadowList(
+		scene::ISceneNode *node, E_SHADOW_MODE shadowMode)
+{
+	m_shadow_node_array.emplace_back(NodeToApply(node, shadowMode));
+}
+
+void ShadowRenderer::removeNodeFromShadowList(scene::ISceneNode *node)
+{
+	for (auto it = m_shadow_node_array.begin(); it != m_shadow_node_array.end();) {
+		if (it->node == node) {
+			it = m_shadow_node_array.erase(it);
+			break;
+		} else {
+			++it;
+		}
+	}
+}
+
+void ShadowRenderer::setClearColor(video::SColor ClearColor)
+{
+	m_clear_color = ClearColor;
+}
+
+void ShadowRenderer::update(video::ITexture *outputTarget)
+{
+	if (!m_shadows_enabled || m_smgr->getActiveCamera() == nullptr) {
+		m_smgr->drawAll();
+		return;
+	}
+
+	if (!shadowMapTextureDynamicObjects) {
+
+		shadowMapTextureDynamicObjects = getSMTexture(
+				std::string("shadow_dynamic_") +
+						itos(m_shadow_map_texture_size),
+				m_texture_format, true);
+	}
+
+	if (!shadowMapClientMap) {
+
+		shadowMapClientMap = getSMTexture(
+				std::string("shadow_clientmap_") +
+						itos(m_shadow_map_texture_size),
+				m_shadow_map_colored ? m_texture_format_color
+						     : m_texture_format,
+				true);
```

Comment:
Applying clang-format made this look _worse_, please undo.

Reply:
Reverted.